File: hestia/etcd/discovery.go
package etcd

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	clientv3 "go.etcd.io/etcd/client/v3"

	"github.com/daheige/hephfx/hestia"
)

var _ hestia.Discovery = (*etcdDiscovery)(nil)

type etcdDiscovery struct {
	client       *clientv3.Client
	prefix       string
	serviceList  map[string][]*hestia.Service
	disableWatch bool // disable watch
	mu           sync.RWMutex
}

// NewDiscovery create a registry interface instance
func NewDiscovery(endpoints []string, opts ...Option) (hestia.Discovery, error) {
	opt := &Options{
		endpoints:    endpoints,
		dialTimeout:  5 * time.Second,
		prefix:       "/hestia/registry-etcd",
		disableWatch: true, // disable watch
	}

	for _, o := range opts {
		o(opt)
	}

	client, err := newEtcdClient(opt)
	if err != nil {
		return nil, err
	}

	e := &etcdDiscovery{
		client:       client,
		prefix:       opt.prefix,
		serviceList:  make(map[string][]*hestia.Service, 20),
		disableWatch: opt.disableWatch,
	}

	e.prefix = strings.TrimPrefix(e.prefix, "/")
	e.prefix = strings.TrimSuffix(e.prefix, "/")
	e.prefix = fmt.Sprintf("/%s", e.prefix) // 格式为/services

	return e, nil
}

// GetServices returns a list of instances
// After we obtain the service instance, we can get the currently available service instance
// from the service list according to different strategies.
func (e *etcdDiscovery) GetServices(ctx context.Context, name string, version string) ([]*hestia.Service, error) {
	var (
		services []*hestia.Service
		exist    bool
	)
	if !e.disableWatch {
		e.mu.RLock()
		services, exist = e.serviceList[name]
		e.mu.RUnlock()
	}

	if !exist {
		var err error
		services, err = e.getServicesByName(ctx, name, version)
		if err != nil {
			return nil, err
		}
		if len(services) == 0 {
			return nil, hestia.ErrServicesNotFound
		}

		e.mu.Lock()
		e.serviceList[name] = services
		e.mu.Unlock()

		if !e.disableWatch {
			go e.watch(context.WithoutCancel(ctx), name, version)
		}
	}

	return services, nil
}

// Get returns an available service instance based on the specified service selection strategy.
// the selection strategy is RoundRobinHandler
func (e *etcdDiscovery) Get(ctx context.Context, name string, version string, strategyHandler ...hestia.StrategyHandler) (*hestia.Service, error) {
	services, err := e.GetServices(ctx, name, version)
	if err != nil {
		return nil, err
	}

	var handler = hestia.RoundRobinHandler
	if len(strategyHandler) > 0 && strategyHandler[0] != nil {
		handler = strategyHandler[0]
	}

	service := handler(services)
	return service, nil
}

// String returns discovery name
func (e *etcdDiscovery) String() string {
	return "etcd"
}

// listen services change
func (e *etcdDiscovery) watch(ctx context.Context, name string, version string) {
	key := e.discoveryKey(name, version)
	e.watchWithCallback(ctx, name, version, func(services []*hestia.Service, err error) {
		if err != nil {
			log.Printf("reload etcd prefix:%s services error:%v", key, err)
			return
		}

		e.mu.Lock()
		e.serviceList[name] = services
		e.mu.Unlock()
	})
}

// watchWithCallback watches the service prefix and invokes callback on every change.
// It is package-private so the etcd resolver can consume real-time updates without
// exposing watch semantics in the public Discovery API.
func (e *etcdDiscovery) watchWithCallback(ctx context.Context, name string, version string,
	callback func([]*hestia.Service, error)) {
	key := e.discoveryKey(name, version)
	// log.Printf("watch etcd prefix:%v\n", key)
	respChan := e.client.Watch(ctx, key, clientv3.WithPrefix())
	for resp := range respChan {
		for _, event := range resp.Events {
			switch event.Type {
			case clientv3.EventTypePut, clientv3.EventTypeDelete:
				services, err := e.getServicesByName(ctx, name, version)
				callback(services, err)
			}
		}
	}
}

func (e *etcdDiscovery) discoveryKey(name, version string) string {
	var key string
	if version == "" {
		key = fmt.Sprintf("%s/%s/", e.prefix, name)
	} else {
		key = fmt.Sprintf("%s/%s/%s/", e.prefix, name, version)
	}

	return key
}

func (e *etcdDiscovery) getServicesByName(ctx context.Context, name string, version string) ([]*hestia.Service, error) {
	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
	defer cancel()

	key := e.discoveryKey(name, version)
	resp, err := e.client.Get(ctx, key, clientv3.WithPrefix())
	if err != nil {
		return nil, err
	}
	if len(resp.Kvs) == 0 {
		return nil, nil
	}

	// 获取所有的服务实例列表
	services := make([]*hestia.Service, 0, len(resp.Kvs))
	for _, kv := range resp.Kvs {
		serviceEntry := &hestia.Service{}
		err = json.Unmarshal(kv.Value, serviceEntry)
		if err != nil {
			log.Printf("unmarshal service failed,error:%v", err)
			continue
		}
		if serviceEntry.Healthy {
			services = append(services, serviceEntry)
		}
	}

	return services, nil
}

File: rs-hestia/src/etcd/discovery.rs
use std::any::Any;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use async_trait::async_trait;
use etcd_client::{Client, EventType, GetOptions, WatchOptions};
use tokio::sync::RwLock;

use crate::Context;
use crate::discovery::Discovery;
use crate::error::{HestiaError, Result};
use crate::service::{Service, StrategyHandler};

use super::{Options, new_etcd_client, normalize_prefix};

/// etcd-backed [`Discovery`] implementation.
#[derive(Clone)]
pub struct EtcdDiscovery {
    client: Client,
    prefix: String,
    disable_watch: bool,
    service_list: Arc<RwLock<HashMap<String, Vec<Service>>>>,
}

/// Creates a new etcd-backed discovery client.
pub async fn new_discovery(options: Options) -> Result<Arc<dyn Discovery>> {
    let client = new_etcd_client(&options).await?;
    let prefix = normalize_prefix(&options.prefix);

    Ok(Arc::new(EtcdDiscovery {
        client,
        prefix,
        disable_watch: options.disable_watch,
        service_list: Arc::new(RwLock::new(HashMap::with_capacity(20))),
    }))
}

#[async_trait]
impl Discovery for EtcdDiscovery {
    async fn get_services(&self, ctx: &Context, name: &str, version: &str) -> Result<Vec<Service>> {
        if !self.disable_watch {
            let list = self.service_list.read().await;
            if let Some(services) = list.get(name) {
                return Ok(services.clone());
            }
        }

        let services = self.get_services_by_name(ctx, name, version).await?;
        if services.is_empty() {
            return Err(HestiaError::ServicesNotFound);
        }

        if !self.disable_watch {
            let mut list = self.service_list.write().await;
            list.insert(name.to_string(), services.clone());

            // Spawn a detached watch so that it survives caller cancellation.
            let this = self.clone();
            let name = name.to_string();
            let version = version.to_string();
            tokio::spawn(async move {
                this.watch(&name, &version).await;
            });
        }

        Ok(services)
    }

    async fn get(
        &self,
        ctx: &Context,
        name: &str,
        version: &str,
        strategy: Option<StrategyHandler>,
    ) -> Result<Service> {
        let services = self.get_services(ctx, name, version).await?;
        let handler = strategy.unwrap_or_else(crate::service::round_robin_handler);
        handler(&services).ok_or(HestiaError::ServicesNotFound)
    }

    fn name(&self) -> &str {
        "etcd"
    }

    fn as_any(&self) -> &dyn Any {
        self
    }
}

impl EtcdDiscovery {
    /// Watches the service prefix and updates the local cache on every change.
    async fn watch(&self, name: &str, version: &str) {
        let name = name.to_string();
        let version = version.to_string();
        self.watch_with_callback(&name, &version, move |_services| {
            // Internal cache update is intentionally a no-op here because
            // `get_services` already writes to the cache. This callback
            // exists so the gRPC resolver can share the same watch path.
        })
        .await;
    }

    /// Watches the service prefix and invokes `callback` on every change.
    ///
    /// This is `pub(crate)` so that the etcd resolver can receive real-time
    /// updates without exposing watch semantics in the public `Discovery` API.
    pub(crate) async fn watch_with_callback(
        &self,
        name: &str,
        version: &str,
        callback: impl Fn(Vec<Service>) + Send + 'static,
    ) {
        let key = discovery_key(&self.prefix, name, version);
        let mut watch_client = self.client.watch_client();
        let mut stream = match watch_client
            .watch(key.as_str(), Some(WatchOptions::new().with_prefix()))
            .await
        {
            Ok(ws) => ws,
            Err(e) => {
                tracing::error!("etcd watch failed: {}", e);
                return;
            }
        };

        loop {
            match stream.message().await {
                Ok(Some(resp)) => {
                    for event in resp.events() {
                        match event.event_type() {
                            EventType::Put | EventType::Delete => {
                                let ctx = Context::new();
                                match self.get_services_by_name(&ctx, name, version).await {
                                    Ok(services) => {
                                        let mut list = self.service_list.write().await;
                                        list.insert(name.to_string(), services.clone());
                                        callback(services);
                                    }
                                    Err(e) => {
                                        tracing::error!(
                                            "reload etcd prefix:{} services error:{}",
                                            key,
                                            e
                                        );
                                    }
                                }
                            }
                        }
                    }
                }
                Ok(None) => break,
                Err(e) => {
                    tracing::error!("etcd watch stream error: {}", e);
                    break;
                }
            }
        }
    }

    async fn get_services_by_name(
        &self,
        _ctx: &Context,
        name: &str,
        version: &str,
    ) -> Result<Vec<Service>> {
        let key = discovery_key(&self.prefix, name, version);
        let mut kv = self.client.kv_client();
        let resp = tokio::time::timeout(
            Duration::from_secs(15),
            kv.get(key.as_str(), Some(GetOptions::new().with_prefix())),
        )
        .await
        .map_err(|e| HestiaError::Other(format!("get services timeout: {}", e)))??;

        let kvs = resp.kvs();
        let mut services = Vec::with_capacity(kvs.len());
        for kv in kvs {
            let entry: Service = match serde_json::from_slice(kv.value()) {
                Ok(s) => s,
                Err(e) => {
                    tracing::warn!("unmarshal service failed, error: {}", e);
                    continue;
                }
            };
            if entry.healthy {
                services.push(entry);
            }
        }

        Ok(services)
    }
}

fn discovery_key(prefix: &str, name: &str, version: &str) -> String {
    if version.is_empty() {
        format!("{}/{}/", prefix, name)
    } else {
        format!("{}/{}/{}/", prefix, name, version)
    }
}
